enhancements/node-healthcheck-operator.md

```diff
@@ -0,0 +1,435 @@
+Node Health Check Enhancement 
+
+---
+title: node-healthcheck-controller
+authors:
+  - "@rgolangh"
+reviewers:
+  - "@abeekof"
+  - "@n1r1"
+  - "@mshitrit"
+  - "@slintes"
+  - "@dhellmann"
+approvers:
+  - "@abeekof"
+  - "@n1r1"
+---
+
+# Node health checking
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [x] Design details are appropriately documented from clear requirements
+- [x] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+
+Enable opt in for automated node health checking and create objects
+for unhealthy nodes, to be handled later by external controllers
+that comply with the existing  [api].
+
+## Motivation
+
+Enable clusters that do not have a functioning Machine API to detect and respond to node-level failures in order to recover workloads , and/or restore cluster capacity.
+
+The existing Machine Health Check controller requires a functioning machine-api. This prevents its use in the majority of Kubernetes and OpenShift clusters.
+The proposed Node Healthcheck Controller (NHC) provides a generic solution that fits all installation types, together with an external remediation API that allows a pluggable remediation system to be installed and remediate unhealthy nodes.
+
+### Goals
+
+- Create an installation-type agnostic node health check solution
+
+- Feature parity with Machine Health Check controller
+
+- Providing a default strategy combined with [poison-pill].
+
+- Endorse usage of external [remediation api]
+
+- Allow defining different unhealthy criteria by node selectors, and have a sane strategy for handling nodes that match multiple NHC CRs.
+
+- Provide defaults health checks and default remedy custom resource creation.
+
+- Long term - Make Machine Health Check controller only implement the external remediation api and drop the default Machine API based remediation mechanism
+
+
+### Non-Goals
+
+- Coordination of remedy operation - operators/controllers implementers
+expected to track the [node maintenance lease proposal]
+
+- Recording long-term stable history of all health-check failures or remediation
+  actions.
+
+- Detection of Machines that never become healthy Nodes is not part of the initial implementation.  Machine API based cluster can/should continue to use MHC at this time.
+
+## Proposal
+The main use-case we target is UPI, user provisioned infrastructure, which do not have
+the Machine-API controlling them. This enables any kind of recovery mechanism to be built and handle remediation.
```

Comment:
I/we should probably stop using UPI as shorthand for "clusters without MAPI" to avoid creating this kind confusion.

To me, why a given cluster doesn't have MAPI or how it got to that state, is less relevant than the number of customers that can't benefit from the existing solution.

```suggestion
The main use-case being targeted is clusters which, for technical and/or political reasons, do not have
the Machine-API controlling them. This enables any kind of recovery mechanism to be built and handle remediation.
```